### ISSUE TITLE
Nested calls to data.table() do not avoid shallow copies

### ISSUE THREAD

**Issue by mattpollock** [Rdatatable/data.table#1629]
For details, please see this [SO question](http://stackoverflow.com/questions/36434717/adding-column-to-nested-r-data-table-by-reference). The gist is that nested `data.table`s can trigger shallow copies when modified, but do so inconsistently depending on how they were constructed. Obviously this presents an issue w.r.t. writing functions that want to modify them. If we can depend on an update be reference code is cleaner. Otherwise recursive functions need to always return stuff and suppress the shallow copy warning messages.

This reproduces it.

```
DT = data.table(d = list(data.table(a=1))); DT$d[[1]][, new_col := NA]
```

Thanks much.

**Comment by jangorecki** [comment]
Thank you for the report, and sorry for such a slow response. The issue has been fixed, probably long time ago. I pushed unit tests to ensure your use case (as described in SO) will continue to work in future.

**Comment by mattdowle** [comment]
Thanks, @mattpollock. I bisected past versions and established this was fixed in v1.12.4 (Oct 2019). Very likely item 14 fixing #3474 : 

14. Subassigning using `$<-` to a `data.table` embedded in a list column of a single-row `data.table` could fail, [#3474](https://github.com/Rdatatable/data.table/issues/3474). Note that `$<-` is not recommended; please use `:=` instead which already worked in this case. Thanks to Jakob Richter for reporting.

PR #4366 adds your example to the test suite. It seems then there was a bug then with `:=` too on such embedded data.table, not just with `$<-`.

**Comment by mattpollock** [comment]
thanks!
